### PR TITLE
Lecture 10 (armdave / Arman Dave)

### PR DESCRIPTION
**Pull request by armdave**

### CODE REVIEW

File: notes/lecture-10-armdave.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+| Title       | Knowledge Distillation                                                                           |
+| ----------- | -------------------------------------------------------------------------------------------------------- |
+| Lecturer    | Song Han                                                                                                 |
+| Date        | 10/13/2022                                                                                               |
+| Note Author | Arman Dave (armdave)                                                                                            |
+| Description | Introduce knowledge distillation, including self and online distillation, as well as distillation for different tasks; introduces network augmentation, a training technique for tiny ML models |
+
+[Slides](https://hanlab.mit.edu/files/course/slides/MIT-TinyML-Lec10-Knowledge-Distillation.pdf)
+
+[Video](https://www.youtube.com/watch?v=tT9Lnt6stwA&feature=youtu.be&ab_channel=MITHANLab)
```
Comment (zhijian-liu) [inline notes/lecture-10-armdave.md:14]:
```suggestion
```

File: notes/lecture-10-armdave.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+- cutout
+
+- mixup
+
+- AutoAugment
+
+2) Dropout (of neurons) during training to avoid overfitting
+
+NetAug: Training Process
```
Comment (zhijian-liu) [inline notes/lecture-10-armdave.md:194]:
```suggestion
```

File: notes/lecture-10-armdave.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+- Step 3: Choose different subnetworks in the augmmented model for auxillary loss
+
+For a tiny NN, NetAug improves both training and val accuracy
+
+For large NN, NetAug improves training but hurts val accuracy
+
+For the right task, NetAug can outpeform KD
+
+NetAug is **orthogonal** to KD
```
Comment (zhijian-liu) [inline notes/lecture-10-armdave.md:216]:
```suggestion
Results:
- For tiny NNs, NetAug improves both training and val accuracy
- For large NNs, NetAug improves training but hurts val accuracy

NetAug is **orthogonal** to KD. For the right task, NetAug can outperform KD. 
```

File: notes/lecture-10-armdave.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+Conventional approach: 
+
+1) Data augmentation during training to avoid overfitting
+
+- cutout
+
+- mixup
+
+- AutoAugment
```
Comment (zhijian-liu) [inline notes/lecture-10-armdave.md:190]: Please add citations for these methods.

File: notes/lecture-10-armdave.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+Beyond classification
+
+### KD for Object Detection
+
+Use different weights for foreground and background classes to handle the class imbalance
+
+How do we convert bounding box regression to a classification problem?
+
+Simply discretize the x and y axis
+
```
Comment (zhijian-liu) [inline notes/lecture-10-armdave.md:163]:
```suggestion
How do we convert bounding box regression to a classification problem? Simply discretize the x and y axis!

```

File: notes/lecture-10-armdave.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+### Self Distillation
+
+Train some "edge" model T
+
+Then create another "edge" model S_1 with same architecture as T and distill T to S_1 in regular teacher -> student model
+
+Then create "edge" model S2, distill S_1 to S_2
+
+Repeat for S_2....S_k
```
Comment (zhijian-liu) [inline notes/lecture-10-armdave.md:128]:
```suggestion
Train some "edge" model $T$

Then create another "edge" model $S_1$ with same architecture as $T$ and distill $T$ to $S_1$ in regular teacher -> student model

Then create "edge" model $S_2$, distill $S_1$ to $S_2$

Repeat for $S_2, ..., S_k$
```

File: notes/lecture-10-armdave.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+* Network accuracy: $T < S_1 < S_2 < ... < S_k$
+
+Best accuracy is actually is an ensemble of all models
+
+### Online Distillation
+
+Aka Deep mutual learning
+
+The teacher model is not pre-trained. Rather, both models are training at the same time
```
Comment (zhijian-liu) [inline notes/lecture-10-armdave.md:138]:
```suggestion
The teacher model is not pre-trained. Rather, both models are trained at the same time
```

File: notes/lecture-10-armdave.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+Solution: **Temperature**
+
+A temperature loss function smooths the probability output
+
+![Temperature](figures/lecture-10/armdave/temperature.png)
+
+**Formal Definition of Knowledge Distillation**
+
+* $p(z_i, T) = \frac{exp(z_i/T)}{\sum{exp(z_j/T)}}$. i,j = 0,1,2...,C-1 where C is the number of classes and T is the temperature (normally set to 1)
```
Comment (zhijian-liu) [inline notes/lecture-10-armdave.md:58]:
```suggestion
* $p(z_i, T) = \frac{exp(z_i/T)}{\sum{exp(z_j/T)}}$, for $i,j = 0,1,2...,C-1$ where C is the number of classes and T is the temperature (normally set to 1)
```

**Review by zhijian-liu — CHANGES_REQUESTED** [review]
Thanks for your contribution! Please address the comments above and re-request the review when it's ready.

File: notes/lecture-10-armdave.md (hunk 4 above)
Comment (armdave) [reply notes/lecture-10-armdave.md]: Done!

**Pull request merged**